subsys/net/lib/lwm2m/ipso_buzzer.c

```diff
@@ -0,0 +1,267 @@
+		return stop_buzzer(&buzzer_data[i], true);
+	} else if (buzzer_data[i].onoff && !buzzer_data[i].active) {
+		return start_buzzer(&buzzer_data[i]);
+	}
+
+	return 0;
+}
+
+static void buzzer_work_cb(struct k_work *work)
+{
+	struct ipso_buzzer_data *buzzer = CONTAINER_OF(work,
+						      struct ipso_buzzer_data,
+						      buzzer_work);
+	stop_buzzer(buzzer, false);
+}
+
+static struct lwm2m_engine_obj_inst *buzzer_create(u16_t obj_inst_id)
+{
+	int index, avail = -1, i = 0, j = 0;
+
+	/* Check that there is no other instance with this ID */
+	for (index = 0; index < ARRAY_SIZE(inst); index++) {
+		if (inst[index].obj && inst[index].obj_inst_id == obj_inst_id) {
+			LOG_ERR("Can not create instance - "
+				"already existing: %u", obj_inst_id);
+			return NULL;
+		}
+
+		/* Save first available slot index */
+		if (avail < 0 && !inst[index].obj) {
+			avail = index;
+		}
+	}
+
+	if (avail < 0) {
+		LOG_ERR("Can not create instance - no more room: %u",
+			obj_inst_id);
+		return NULL;
+	}
+
+	/* Set default values */
+	k_delayed_work_init(&buzzer_data[avail].buzzer_work, buzzer_work_cb);
+	buzzer_data[avail].level.val1 = 50; /* 50% */
+	buzzer_data[avail].level.val2 = 0;
+	buzzer_data[avail].delay_duration.val1 = 1; /* 1 seconds */
+	buzzer_data[avail].delay_duration.val2 = 0;
+	buzzer_data[avail].min_off_time.val1 = 0;
+	buzzer_data[avail].min_off_time.val2 = 0;
+	buzzer_data[avail].trigger_offset = 0U;
+	buzzer_data[avail].onoff = false;
+	buzzer_data[avail].active = false;
+	buzzer_data[avail].obj_inst_id = obj_inst_id;
+
+	init_res_instance(res_inst[avail], ARRAY_SIZE(res_inst[avail]));
+
+	/* initialize instance resource data */
+	INIT_OBJ_RES(BUZZER_ON_OFF_ID, res[avail], i,
+		     res_inst[avail], j, 1, true,
+		     &buzzer_data[avail].onoff,
+		     sizeof(buzzer_data[avail].onoff),
+		     NULL, NULL, onoff_post_write_cb, NULL);
+	INIT_OBJ_RES_DATA(BUZZER_LEVEL_ID, res[avail], i, res_inst[avail], j,
+			  &buzzer_data[avail].level,
+			  sizeof(buzzer_data[avail].level));
+	INIT_OBJ_RES_DATA(BUZZER_DELAY_DURATION_ID, res[avail], i,
+			  res_inst[avail], j,
+			  &buzzer_data[avail].delay_duration,
+			  sizeof(buzzer_data[avail].delay_duration));
+	INIT_OBJ_RES_DATA(BUZZER_MINIMUM_OFF_TIME_ID, res[avail], i,
+			  res_inst[avail], j,
+			  &buzzer_data[avail].min_off_time,
+			  sizeof(buzzer_data[avail].min_off_time));
+	INIT_OBJ_RES_OPTDATA(BUZZER_APPLICATION_TYPE_ID, res[avail], i,
+			     res_inst[avail], j);
+	INIT_OBJ_RES_DATA(BUZZER_DIGITAL_STATE_ID, res[avail], i,
+			  res_inst[avail], j,
+			  &buzzer_data[avail].active,
+			  sizeof(buzzer_data[avail].active));
+
+	inst[avail].resources = res[avail];
+	inst[avail].resource_count = i;
+
+	LOG_DBG("Create IPSO Buzzer instance: %d", obj_inst_id);
+
+	return &inst[avail];
+}
+
+static int ipso_buzzer_init(struct device *dev)
+{
+	/* Set default values */
+	(void)memset(inst, 0, sizeof(*inst) * ARRAY_SIZE(inst));
+	(void)memset(res, 0, sizeof(struct lwm2m_engine_res) *
```

Comment:
These two memset() seems to be useless as the variables are static.

Reply:
You are correct!  And this was copy/pasted into every single LwM2M object that is in Zephyr.  I'm going to do 2 things here:
1. Remove these memset lines from the *_init() functions of the new objects being introduced.
2. Add a memset line to the *_create() function above here to make sure the resource data that the new object instance is using is reset (it could have been used prior and then freed up again via DELETE op)
3. Stage a patch at the end of this series doing the same for all of the existing objects.